From 27b8312569d4c3f8dbf527d04b47690be641eb17 Mon Sep 17 00:00:00 2001
From: Gunmeet Singh <gunmeet.singhx.chadha@intel.com>
Date: Tue, 6 Oct 2015 14:11:20 +0530
Subject: [PATCH] ASoC: codec: rt5670: Fix for channel swapping issue.

The following codec driver patch will fix the L and R channel
swapping issue when audio is played on Speakers.

Change-Id: I4865bf1156e423c6be02f415ec3ecf7ee97232dd
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-449
Signed-off-by: Gunmeet Singh <gunmeet.singhx.chadha@intel.com>
Reviewed-on: https://android.intel.com:443/424499
(cherry picked from commit 316e9e40f7f41614244595d0ddd9ffa396b48628)
---
 sound/soc/codecs/rt5670.c |  8 ++++++++
 sound/soc/codecs/rt5670.h | 10 ++++++++++
 2 files changed, 18 insertions(+)

diff --git a/sound/soc/codecs/rt5670.c b/sound/soc/codecs/rt5670.c
index 05bb069..adeb4d9 100644
--- a/sound/soc/codecs/rt5670.c
+++ b/sound/soc/codecs/rt5670.c
@@ -3525,6 +3525,14 @@ static int rt5670_probe(struct snd_soc_codec *codec)
 	}
 
 	rt5670_reg_init(codec);
+
+	/* Swap SPK L/R channel if codec version is more than 4 */
+	if ((snd_soc_read(codec, RT5670_VENDOR_ID) & 0xff) >= 4)
+		snd_soc_update_bits(codec, RT5670_GPIO_CTRL3,
+			RT5670_GP9_OUT_MASK | RT5670_GP8_OUT_MASK,
+			RT5670_GP9_OUT_LO | RT5670_GP8_OUT_HI);
+
+
 #ifdef JD1_FUNC
 	snd_soc_update_bits(codec, RT5670_PWR_ANLG1,
 			RT5670_PWR_MB | RT5670_PWR_BG,
diff --git a/sound/soc/codecs/rt5670.h b/sound/soc/codecs/rt5670.h
index 9666e66..e7f52ab 100644
--- a/sound/soc/codecs/rt5670.h
+++ b/sound/soc/codecs/rt5670.h
@@ -1616,6 +1616,16 @@
 #define RT5670_GP1_P_NOR			(0x0)
 #define RT5670_GP1_P_INV			(0x1)
 
+/* GPIO Control 3 (0xc2) */
+#define RT5670_GP9_OUT_MASK			(0x1 << 10)
+#define RT5670_GP9_OUT_SFT			10
+#define RT5670_GP9_OUT_LO			(0x0 << 10)
+#define RT5670_GP9_OUT_HI			(0x1 << 10)
+#define RT5670_GP8_OUT_MASK			(0x1 << 7)
+#define RT5670_GP8_OUT_SFT			7
+#define RT5670_GP8_OUT_LO			(0x0 << 7)
+#define RT5670_GP8_OUT_HI			(0x1 << 7)
+
 /* Scramble Function (0xcd) */
 #define RT5670_SCB_KEY_MASK			(0xff)
 #define RT5670_SCB_KEY_SFT			0
-- 
1.9.1

